### DISCUSSION TITLE
Cropping by four percent values

### DISCUSSION THREAD

**Discussion by speedgeil** [ImageMagick/ImageMagick#4695] [Digital Image Processing]
Hi together,
I hope this is the right category.
This cropping topic is driving me nuts. I have four known crop percent values. They represent the amount of crop from the edges of an image to inwards, like  

Top 10%, 
Left 20%, 
Right 30%,
Bottom 40%

I know this should be possible with -gravity but i just don't get the instructions from the user guide, mostly because of missing examples i think. Could someone please help me out with the right code snippet before my brain pulls the plug?

**Comment by fmw42** [comment]
ImageMagick crop uses WxH+X+Y.  WxH can be % but X and Y must be pixels. So W=100-left-right=100-20-30=50%, H=100-top-bottom=100-10-40=50%.  X=(left%/100)\*input width in pixels, Y=(top%/100)\*input height in pixels.  So you have find the Input Width and Input Height. This can be done in IM 7 in one command line, but in IM 6, you would have find the input width and height in separate commands and save as variables.

What is your IM version and platform/OS?

**Reply by speedgeil** [reply]
Hi Fred,
thanks for your quick answer. I'm on Windows 11 and IM V7.

**Reply by fmw42** [reply]
Then this should work.

`magick image.suffix -crop "%[fx:100-20-30]%x[fx:100-10-40]%+[fx:w*20/100]+[fx:h*10/100] cropped.suffix`

w and h are fx variables that find the input width and height automatically

Let me know if that works for you.  If in a .bat file, then the % need to be doubled to %%.

Sorry, I am not a Windows user. I use a Mac. So I do not know to put variables for your values into the command. A Windows user will have help if you do not know.

**Reply by fmw42** [reply]
Alternately, you can convert all your % to pixels, by multiplying the fractions by the input width and input height. Then do the same operation, but leave off the % and use the actual w or h for 100 in the formula.


```
T=height*left%/100
B=height*bottom%/100
L=width*left%/100
R=width*right%/100
magick image.suffix -crop "%[fx:w-L-R]x[fx:h-T-B]%+L+T cropped.suffix

```

or

```
T=height*left%/100
B=height*bottom%/100
L=width*left%/100
R=width*right%/100
W=input_width-L-R
H=input_height-T-B
magick image.suffix -crop WxH+L+T cropped.suffix

```

**Reply by speedgeil** [reply]
Thank you Fred!
You know i'm working with node.js child_process in the backgrounf but i did not want to complicate my question too much.
Ok, so i put your code and mixed it with my variables for crops etc... The var cmd is later passed to the child_process exec. Well, i did not get it to work. Maybe it's just a typo? The percentages for crops are fine, i can log them in console. I think i have to capture command line error out. Or do you see my mistake?
`  var cmd = "magick" + '"' + inputPath + '"' + " image.suffix -crop %[fx:100-" + cropLeft + "-" + cropRight + "]%x[fx:100-" + cropTop + "-" + cropBottom + "]%+[fx:w*20/100]+[fx:h*10/100] cropped.suffix -channel All -contrast-stretch 0.4x0.3% " + '"' + outputPath + ".tif" + '"'
`

**Reply by fmw42** [reply]
you have to substitute your actual input image file for image.suffix. Also you have cropped.suffix and then you specify the output. In my command cropped.suffix was the output. So you would have had to replace that with your actual output file. In your command you will have to remove cropped.suffix since you subsequently add more commands to the command line and then specify an output path and suffix.  Other than that, I do not know node.js syntax.

**Reply by kramsurfer** [reply]
I believe you need double %'s in your string you are passing to CMD.   Make all your % symbols %%

https://ss64.com/nt/syntax-percent.html

**Reply by speedgeil** [reply]
Yes you are right. I apologize, as i don't really know what i'm doing.
I captured stderr and it confirms. I'll fix that and get back once i got it working (or run into the next problem)

convert: UnableToOpenBlob 'image.suffix': No such file or directory @ error/blob.c/OpenBlob/3527.
convert: UnableToOpenConfigureFile `delegates.xml' @ warning/configure.c/GetConfigureOptions/705.
convert: UnableToOpenBlob 'image.suffix': No such file or directory @ error/blob.c/OpenBlob/3527.
convert: NoDecodeDelegateForThisImageFormat `SUFFIX' @ error/constitute.c/ReadImage/572.
convert: InvalidArgument '-crop': %[fx:100-2-96]%x[fx:100-4-94]%+[fx:w*20/100]+[fx:h*10/100] @ error/convert.c/ConvertImageCommand/1229.

**Reply by speedgeil** [reply]
> I believe you need double %'s in your string you are passing to CMD. Make all your % symbols %%
> 
> https://ss64.com/nt/syntax-percent.html

Thank you Kramsurfer. It's actually not needed when working with child_process.

**Reply by speedgeil** [reply]
> -crop "%[fx:100-20-30]%x[fx:100-10-40]%+[fx:w*20/100]+[fx:h*10/100] cropped.suffix

Hm! It doesn't work. The stderr output complains about code the looks correct to me.
What am i doing wrong?
`InvalidArgument '-crop': %[fx:100-4-4]%x[fx:100-4-4]%+[fx:w*20/100]+[fx:h*10/100]`

**Reply by speedgeil** [reply]
> you have to substitute your actual input image file for image.suffix. Also you have cropped.suffix and then you specify the output. In my command cropped.suffix was the output. So you would have had to replace that with your actual output file. In your command you will have to remove cropped.suffix since you subsequently add more commands to the command line and then specify an output path and suffix. Other than that, I do not know node.js syntax.

Ok, i updatet IM to the latest version and now i get this:
`magick: MissingArgument `-crop' at CLI arg 2 @ fatal/magick-cli.c/ProcessCommandOptions/447`

This is my code:
`var cmd = "magick " + '"' + inputPath + '"' + " -crop " + "\"%[fx:100-" + cropLeft + "-" + (100-cropRight) + "]%x[fx:100-" + cropTop + "-" + (100-cropBottom) + "]%+[fx:w*20/100]+[fx:h*10/100] " + '"' + outputPath + ".tif" + '"'`

**Reply by fmw42** [reply]
Sorry, but I do not understand node.js.  Why don't you just try the simple command without all the argument passing until you get that to work.

**Reply by speedgeil** [reply]
> -crop "%[fx:100-20-30]%x[fx:100-10-40]%+[fx:w*20/100]+[fx:h*10/100]

I just did. Same result:
`magick: MissingArgument `-crop' at CLI arg 2 @ fatal/magick-cli.c/ProcessCommandOptions/447.

(anonymous) @ main.js:18
exithandler @ node:child_process:333
emit @ node:events:378
maybeClose @ node:internal/child_process:1067
ChildProcess._handle.onexit @ node:internal/child_process:301`

Just as a sidenote: Every other code worked so far.

**Reply by fmw42** [reply]
You need ending quotes on the crop arguments. You have a leading quote and need and ending one.

**Reply by speedgeil** [reply]
> You need ending quotes on the crop arguments. You have a leading quote and need and ending one.

Please have a look at your code from your first answer. I just copied what you gave me. But unfortunately it won`t work even with leading an closing quotes. Not even directly in the command line. Could you please review your code and confirm it working?

**Comment by snibgo** [comment]
Deleted.

**Comment by fmw42** [comment]
You need a multiply symbol between w and 20 and also between h and 10. So change

`-crop "%[fx:100-20-30]%x%[fx:100-10-40]%+%[fx:w20/100]+%[fx:h10/100] cropped.suffix
`

to

`-crop "%[fx:100-20-30]%x%[fx:100-10-40]%+%[fx:w*20/100]+%[fx:h*10/100]" cropped.suffix
`

**Reply by kramsurfer** [reply]
What's the double quote before the first % there for?  I don't see a pairing for it.

**Reply by speedgeil** [reply]
> What's the double quote before the first % there for? I don't see a pairing for it.

I think Fred knows. Had a hard time escaping it in js :-D

**Reply by fmw42** [reply]
Double quotes are needed before and after the crop arguments if spaces are included

so change

`-crop "%[fx:100-20-30]%x[fx:100-10-40]%+[fx:w*20/100]+[fx:h*10/100] cropped.suffix
`
to

`-crop "%[fx:100-20-30]%x[fx:100-10-40]%+[fx:w*20/100]+[fx:h*10/100]" cropped.suffix`

**Reply by speedgeil** [reply]
> Double quotes are needed before and after the crop arguments if spaces are included
> 
> so change
> 
> `-crop "%[fx:100-20-30]%x[fx:100-10-40]%+[fx:w*20/100]+[fx:h*10/100] cropped.suffix ` to
> 
> `-crop "%[fx:100-20-30]%x[fx:100-10-40]%+[fx:w*20/100]+[fx:h*10/100]" cropped.suffix`

That produces the same error. Just to make clear, i used the windows command line tool instead of node.js: 
`magick C:\Users\speedgeil\Pictures\input.dng -crop "%[fx:100-2-4]%x[fx:100-4-6]%+[fx:w*20/100]+[fx:h*10/100]" C:\Users\speed\Pictures\output.tif`

Error:
`magick: MissingArgument `-crop' at CLI arg 2 @ fatal/magick-cli.c/ProcessCommandOptions/447.`

**Reply by kramsurfer** [reply]
Or remove the leading quote, if you've no spaces in your crop string.  Which is probably a better input for IM.

**Reply by speedgeil** [reply]
> Or remove the leading quote, if you've no spaces in your crop string. Which is probably a better input for IM.

tried that too with the same result. Could someone please try that snippet if it works at all?

**Reply by fmw42** [reply]
Is this in a .bat file or in the CMD window? If in a .bat file, you need to double the % to %%

**Reply by fmw42** [reply]
Change

`-crop "%[fx:100-2-4]%x[fx:100-4-6]%+[fx:w*20/100]+[fx:h*10/100]"`

to

`-crop "%[fx:100-2-4]%x%[fx:100-4-6]%+%[fx:w*20/100]+%[fx:h*10/100]"`

The % should be after the x and before the [fx:...

**Reply by kramsurfer** [reply]
You need % signs in front of each bracketed var set 

-crop %[fx:100-20-30]x%[fx:100-10-40]+%[fx:w*20/100]+%[fx:h*10/100] cropped.suffix

Dang, I was beat.. hahaha

**Comment by fmw42** [comment]
I had a typo.  Sorry.

Try

`-crop "%[fx:100-20-30]%x%[fx:100-10-40]%+%[fx:w*20/100]+%[fx:h*10/100]" cropped.suffix
`

it is always %[fx:....]  then % at the end for specifying width and or height by percent.  The leading % is part of the syntax. The trailing % is if you want the units to be percent.

**Reply by speedgeil** [reply]
> -crop "%[fx:100-20-30]%x%[fx:100-10-40]%+%[fx:w*20/100]+%[fx:h*10/100] cropped.suffix

Is the leading quote "% correct without ending quote?

**Reply by fmw42** [reply]
No, either no quotes and no spaces or quotes at beginning and ending of crop arguments

**Reply by speedgeil** [reply]
> -crop "%[fx:100-20-30]%x%[fx:100-10-40]%+%[fx:w*20/100]+%[fx:h*10/100]" cropped.suffix

Ok, at least it does something now. But it will crop the image to half, no matter what the values are. stderr is giving me this
stderr: magick: UnknownImageProperty "%+" @ warning/property.c/InterpretImageProperties/3762.

**Reply by fmw42** [reply]
Please do not use "quoting" for command, since it messes up syntax for multiplication symbol *.  Please use the code button <> rather than the quote button. So please check that your command has w\*20/100 and h\*10/100

**Reply by fmw42** [reply]
Try using just one percent after the [fx:....].  IM will know that both W and H are to be in percents if just one % is found

`-crop "%[fx:100-20-30]%x%[fx:100-10-40]+%[fx:w*20/100]+%[fx:h*10/100]" cropped.suffix
`

or without double quotes

`-crop %[fx:100-20-30]%x%[fx:100-10-40]+%[fx:w*20/100]+%[fx:h*10/100] cropped.suffix
`

**Reply by speedgeil** [reply]
I just quoted your comment. Nothing else. The command has ` w*20/100 and h*10/100` but does not work. Are you sure you checked your code for typos before posting?

**Reply by speedgeil** [reply]
I used `-crop %[fx:100-20-30]%x%[fx:100-10-40]+%[fx:w*20/100]+%[fx:h*10/100] cropped.suffix` in the command line and it still cuts the image down to 50% crop. Even when i change the values to `-crop %[fx:100-5-5]%x%[fx:100-5-5]+%[fx:w*20/100]+%[fx:h*10/100] cropped.suffix`

**Reply by fmw42** [reply]
What code of mine?  I don't have w20/100, I had w\*20/10.  This forum messes up multiplication symbol *, if not put in <> code or escaped

**Reply by speedgeil** [reply]
Just the last code i quoted. It does not work.

**Reply by fmw42** [reply]
Your quoted code is false, because it is missing the * for multiplication. I did not write it without the *. The system loses the * when you do not put your code in code format using the <> button!

**Comment by fmw42** [comment]
Looks like the percents need to be escaped, at least on Unix

`magick lena.jpg -crop "%[fx:100-5-5]\%x%[fx:100-5-5]\%+%[fx:w*20/100]+%[fx:h*10/100]" x.jpg`

This works for me.

So on Windows, the escape is ^ rather than \, so try

`magick lena.jpg -crop "%[fx:100-5-5]^%x%[fx:100-5-5]^%+%[fx:w*20/100]+%[fx:h*10/100]" x.jpg`

**Reply by speedgeil** [reply]
No when i run the code stdout is empty, what means no error occured. As the image is cropped to half i suspect the logic of the code to be wrong. After running with escaped percents `-crop "%[fx:100-5-5]^%x%[fx:100-5-5]^%+%[fx:w*20/100]+%[fx:h*10/100]"`
I get an stderr with 
magick: MissingArgument `-crop' at CLI arg 2 @ fatal/magick-cli.c/ProcessCommandOptions/447.
again. Percents need to be escaped with %% only in bat files on windows, but not in the command line

**Reply by fmw42** [reply]
My code works in Unix with escaped %.  Sorry, I am not Windows user so do not know what you need to do to escape them. One of the Windows users will need to help you further.

Perhaps try without the quotes.

**Reply by fmw42** [reply]
Or try

`-crop "%[fx:100-5-5]%%x%[fx:100-5-5]%%+%[fx:w*20/100]+%[fx:h*10/100]"`

**Comment by fmw42** [comment]
try quotes around each %[fx: ... ]

`-crop "%[fx:100-5-5]"%x"%[fx:100-5-5]"%+"%[fx:w*20/100]"+"%[fx:h*10/100]"`

**Reply by speedgeil** [reply]
`-crop "%[fx:100-5-5]%%x%[fx:100-5-5]%%+%[fx:w*20/100]+%[fx:h*10/100]"`

will crop portrait oriented images on the top, left and bottom and landscape images on the left and top

**Reply by speedgeil** [reply]
`-crop "%[fx:100-5-5]"%x"%[fx:100-5-5]"%+"%[fx:w*20/100]"+"%[fx:h*10/100]"`
will do nothing to landscape oriented images and will crop portraits lower third

**Reply by fmw42** [reply]
without specifying -gravity, -crop will crop from top left corner by the arguments provided. I have no idea what is going on in your Windows syntax. Sorry. One of the Windows users will have to help. What you are saying does not make sense to me for those formats. Is your input JPG? If so, you may need to use -auto-orient to correct for proper camera orientation.

**Reply by speedgeil** [reply]
I'm using dng. Could you explain what `"%[fx:w*20/100]"+"%[fx:h*10/100]` is supposed to do?

**Reply by fmw42** [reply]
Lets get rid of the percentages. Try

`-crop "%[fx:w*(100-5-5)/100]x%[fx:h*(100-5-5)/100]+%[fx:w*20/100]+%[fx:h*10/100]"`

Does that work?

**Reply by fmw42** [reply]
`"%[fx:w*20/100]"+"%[fx:h*10/100]`  are the +X and +Y offsets from the top left corner for the crop in the format of WxH+X+Y. So one starts your crop at that location and then gets the specified W and H for the crop size from the other arguments. We have just converted your left and top from percents here to pixels. w and h will get the actual input width and height as arguments to fx:

**Reply by speedgeil** [reply]
Thanks for explaining.

`-crop "%[fx:w*(100-5-5)/100]x%[fx:h*(100-5-5)/100]+%[fx:w*20/100]+%[fx:h*10/100]"`
will crop top, left and bottom on portraits and left and top on landscapes

**Reply by fmw42** [reply]
I do not understand what you are saying about portrait and landscape. It should crop from top left in both cases, unless your images are rotated and you have not used -auto-orient. That could mess you up. Try adding -auto-orient before the crop.

**Reply by speedgeil** [reply]
I already added auto-orient but it does nothing. Images are not rotated or anything.

**Reply by fmw42** [reply]
`-crop "%[fx:w*(100-5-5)/100]x%[fx:h*(100-5-5)/100]+%[fx:w*20/100]+%[fx:h*10/100]"`

> will crop top, left and bottom on portraits and left and top on landscapes

I do not understand your comment.  Bottom has nothing to do with IM crop. It only starts at top/left and that is the same left/top.

**Reply by speedgeil** [reply]
It is what it is. That is how the images are returned and i described how they were cropped. Even the amount of cropping is wrong. It's always much more than 5%

**Reply by fmw42** [reply]
Perhaps this will help you understand.

![crop_diagram](https://user-images.githubusercontent.com/13458218/148876264-abf0b621-a0be-45eb-99b5-1c0d7145996d.png)

So in pixel units for left, right, top, bottom:

```
W=input width - left - right
H=input height - top - bottom
X=left
Y=top

```

or in percents for W and H and pixels for left and right:

```
W=(100 - left - right)%
H=(100 - top - bottom)%
X=left
Y=top

```
If all values are percents, then we just convert them to pixels:

```
W=(input width)*(100 - left - right)
H=(input height)*(100 - top - bottom)
X=(Input width)*left/100
Y=(Input height)*top/100

```

I do not know what to tell you about your crop results. Perhaps you should post and example input that does not work, your code and your resulting image, so we can check it all out.

**Reply by kramsurfer** [reply]
Something like this:  ( wrote in browser here, not tested )
` + " -crop " + "\"%%[fx:100-" + ( cropLeft + cropRight ) + "]x%%[fx:100-" + ( cropTop + cropBottom ) + "]+%%[fx:w*20/100]+%%[fx:h*10/100] " +
`

**Reply by speedgeil** [reply]
Thank you kramsurfer. It has the same effect as Fred's scripts. Maybe it has to do with the square brackets? Can't find such syntax in the documentation.

**Reply by speedgeil** [reply]
Fred: Thanks for your help. All i can say is that it doesn't matter what file type i use with your code or if i execute it as bat/command line or node.js... It's always the same result. Did you test the code yourself on Unix?

**Reply by fmw42** [reply]
Yes, the code I wrote above with the escaped % after the square brackets works fine on my Mac. That is Unix syntax. My understanding is that in Windows you would need %% after the square brackets. But I am not a Windows users so do not know the idiosyncrasies of Windows.

try

`-crop "%[fx:100-5-5]"%%x"%[fx:100-5-5]"%%+"%[fx:w*20/100]"+"%[fx:h*10/100]"`

**Reply by fmw42** [reply]
Perhaps user snibgo can set this straight. He is the expert on Windows syntax for IM.

**Reply by speedgeil** [reply]
Fred, the double percent are not needed when running the command from windows command line (same as terminal on Mac). You would need them when making a batch file (.bat). Node.js doesn't need them aswell. Node.js works across Unix and Windows and will work with the standard syntax from imagemagick's documentation.

**Reply by fmw42** [reply]
Please see https://legacy.imagemagick.org/Usage/windows/

**Reply by kramsurfer** [reply]
Good to know.  I typically create a .bat file and then execute that. Adding a pause to the bat allows for debugging so I can easily see any issues.

**Comment by snibgo** [comment]
There seems to be a bug when using both "%[fx:...]" and suffix "%" (the suffix means "these numbers are percentages of width and height"). Here are command-line examples (so "%" must NOT be doubled). toes.png is 267x233 pixels.

```
f:\web\im>%IMG7%magick toes.png -crop 10x20%+3+4 info:
toes.png PNG 27x47 267x233+3+4 16-bit sRGB 320268B 0.000u 0:00.002   <== correct result

f:\web\im>%IMG7%magick toes.png -crop %[fx:10]x20%+3+4 info:
toes.png PNG 27x47 267x233+3+4 16-bit sRGB 320268B 0.000u 0:00.001   <== correct result

f:\web\im>%IMG7%magick toes.png -crop %[fx:10]x%[fx:20]%+3+4 info:
magick: unknown image property "%+" @ warning/property.c/InterpretImageProperties/3765.
toes.png PNG 10x203 267x233+4+0 16-bit sRGB 320268B 0.000u 0:00.002   <== WRONG result
```

Perhaps this is a bug only on Windows. I don't know.

Anyhow, I don't use both "%[fx:...]" and suffix "%". We don't need the "%" suffix if we multiply the "%[fx:...]" expressions.
```
f:\web\im>%FXOLD%magick toes.png -crop %[fx:w*0.10]x%[fx:h*0.20]+3+4 info:
toes.png PNG 27x47 267x233+3+4 16-bit sRGB 320268B 0.688u 0:00.679   <== correct result
```

Does that help? I don't use Node.js.

**Reply by fmw42** [reply]
snibgo, did you ever try some kind of escape on the trailing %?

**Comment by fmw42** [comment]
Did you ever try converting all the percents to pixels as I wrote a while back?

`-crop "%[fx:w*(100-5-5)/100]x%[fx:h*(100-5-5)/100]+%[fx:w*20/100]+%[fx:h*10/100]"`

**Comment by GeeMack** [comment]
Regarding the use of percent signs in Windows 10 BAT scripts, I've used several scripts since the beginning of IMv7 with these and nearly identical operations, all with double percent signs at the end...

    -resize 100x%%[fx:t?100:rand()*50+50]%%
    ...
    -roll +0+%%[fx:rand()*50-25]%%
    ...
    -scale 100x%%[fx:rand()*10+90]%%
    ...
    -evaluate set %%[fx:rand()*30+70]%%

As long as there aren't greater-than, less-than, or pipes in the expressions they work in most cases with or without quotes. I don't know why it might be different from system to system, but when I want to express an IM value as a percent, I use two percent signs at the end and everything works as expected.

But as **fmw42** mentioned, some pretty basic FX expressions can be used to avoid the use of those trailing percent signs. To take an input image and remove 10% from the top, 40% from the bottom, 20% from the left side, and 30% from the right, I'd probably work with a command like this..

    magick input.png -crop %[fx:w-(w*0.2)-(w*0.3)]x%[fx:h-(h*0.1)-(h*0.4)]+%[fx:w*0.2]+%[fx:h*0.1] output.png

**Reply by fmw42** [reply]
> -evaluate set %%[fx:rand()*30+70]%%

So it looks like the % on both sides of the square bracket need to be doubled. Is that correct? So Perhaps this will work for him

`-crop %%[fx:100-5-5]%%x%%[fx:100-5-5]%%+%%[fx:w*20/100]+%%[fx:h*10/100]`

I would be curious to know if that works, though converting W and H from percents to pixels as above is likely the safest. But then am I correct in that it would need to be

`-crop %%[fx:w*(100-5-5)/100]x%%[fx:h*(100-5-5)/100]+%%[fx:w*20/100]+%%[fx:h*10/100]`

**Comment by speedgeil** [comment]
Sorry for the late reply. I tried all your suggestions. None of them worked. 
What works for landscape oriented image is the simple: `-crop " + croppedDim + "+" + leftPos + "+" + topPos + "` or without variables: `-crop 3000x2000+150+150`
But for portrait oriented images they are still horizontally cut to nearly half. -orient does not change anything about that. I double checked the results of my variables in console. I'm sure i made a mistake somewhere but i don't see it. It seems like IM always treats the image horizontally. Weird!

**Reply by fmw42** [reply]
Did you test the commands that I suggested simply in CMD window and not in node.js.  You complicate things testing there before doing it simply in CMD window.

**Reply by speedgeil** [reply]
> Did you test the commands that I suggested simply in CMD window and not in node.js. You complicate things testing there before doing it simply in CMD window.

I do that the whole time

**Comment by snibgo** [comment]
This IM bug shows up at the Windows CMD console. Doubling the % for BAT files is a different issue that isn't an IM bug, but a Windows "feature".

The error message from the bug is:
```
unknown image property "%+" 
```
... so we know that IM can see the %.

Having said that, doubling the suffix % does "fix" the problem. IM then sees two % signs, which it object to, but doesn't. I suspect that if this bug is ever fixed, doubling the suffix % will raise an error.

In a BAT file we need to double each %, so if we use this fix in a BAT file, we need four of them in that position:
```
magick -debug command toes.png -crop %%[fx:10]x%%[fx:20]%%%%+3+4 info:
```

> Did you ever try converting all the percents to pixels as I wrote a while back?

Yes, as I showed in my post.

> So it looks like the % on both sides of the square bracket need to be doubled. Is that correct?

No, at the CMD command line, that is not correct. Doubling the % that occurs in "%[fx:" raises an error, as it should, although the error message is misleading:
```
invalid argument for option '-crop' '%[fx:10]x20%+3+4'
```

@speedgeil wrote:
> But for portrait oriented images they are still horizontally cut to nearly half. -orient does not change anything about that.

Please post a sample "portrait" image. I suspect it is really a "horizontal" image, with metadata that says it should be rotated.

If you can't post a sample image, than what is the text output from:
```
magick yourfile.ext info:
```
I suspect the width will be larger than the height.

**Reply by fmw42** [reply]
My question about testing was directed to the OP.  I also suggested he post and example that fails, but he has not done so, so far. The %% was suggested by GeeMack, if I understood correctly.

**Reply by speedgeil** [reply]
> magick yourfile.ext info:

this is what `info:` gives me: ` DNG 3588x5172 5172x3588+0+0 16-bit sRGB 65.4563MiB 0.000u 0:00.000`

**Comment by snibgo** [comment]
Another workaround for this bug is to put the suffix % at the end of the crop string. Windows CMD console format:
```

magick toes.png -crop "%[fx:10]x%[fx:20]+3+4%" info:
toes.png PNG 27x47 267x233+3+4 16-bit sRGB 320268B 0.844u 0:00.844   <== correct result
```
I dislike this workaround because we might think the offsets +3 and +4 are also percentages.

**Reply by speedgeil** [reply]
> Another workaround for this bug is to put the suffix % at the end of the crop string. Windows CMD console format:
> 
> ```
> 
> magick toes.png -crop "%[fx:10]x%[fx:20]+3+4%" info:
> toes.png PNG 27x47 267x233+3+4 16-bit sRGB 320268B 0.844u 0:00.844   <== correct result
> ```
> 
> I dislike this workaround because we might think the offsets +3 and +4 are also percentages.

At the moment i don't understand why we are dealing with the "percent bug" as we easily could convert to decimals: 100% = 1 
Or does fx strictly need percent?

**Comment by snibgo** [comment]
> 3588x5172 5172x3588+0+0

So your image is 3588x5172 "portrait" but on a 5172x3588 "horizontal" canvas. I suggest you insert "+repage" as your first operation.

> Or does fx strictly need percent?

No, it doesn't. My recommendation is not to use percent.
